File: C++/Programmers/Hash/phone_list.cpp
/*
https://programmers.co.kr/learn/courses/30/parts/12077
PROGRAMMERS HASH Phone Book
*/

#include <string>
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

bool solution(vector<string> phone_book) {
    sort(phone_book.begin(), phone_book.end());

    for(int i=0; i< phone_book.size()-1; i++)
        if(phone_book[i] == phone_book[i+1].substr(0, phone_book[i].size())) return false;
    
    return true;
}

/*
// 히든테케 다 통과하는데 효율성 반만 통과해서 구글링한 결과 길이 순으로 정렬하지 않고, 사전 순으로 정렬하면 인접 인덱스만 비교하면 된다는 것을 알게되었다.
bool compare (string a, string b) {
    return a.size() < b.size();
}
bool solution(vector<string> phone_book) {
    sort(phone_book.begin(), phone_book.end(), compare);
    
    for (int i=0;i<phone_book.size()-1;i++){
        for(int j=i+1;j<phone_book.size();j++){
            if (phone_book[i] == phone_book[j].substr(0, phone_book[i].size()))
                return false;
        }
    }
  
    return true;
}*/

int main(void){
    vector<string> phone_book1{"119", "97674223", "1195524421"};
    vector<string> phone_book2{"123", "456", "789"};
    vector<string> phone_book3{"12", "123", "1235", "567", "88"};
    vector<string> phone_book4{"113333", "115555", "345555", "555555", "345444"};

    if(solution(phone_book1)==false) cout<<"RIGHT1"<<endl;
    if(solution(phone_book2)==true) cout<<"RIGHT2"<<endl;
    if(solution(phone_book3)==false) cout<<"RIGHT3"<<endl;
    if(solution(phone_book4)==true) cout<<"RIGHT4"<<endl;
}

File: Python/Algorithm/hash/phone_list.py
"""
https://programmers.co.kr/learn/courses/30/parts/12077
PROGRAMMERS HASH Phone Book
"""

def solution(phone_book):
    phone_book.sort()
    for i in range(len(phone_book)-1):
        len_index = len(phone_book[i])
        if phone_book[i] in phone_book[i+1][:len_index]:
            return False
    return True

def main():
    phone_book1 = ["119", "97674223", "1195524421"]
    phone_book2 = ["123", "456", "789"]
    phone_book3 = ["12", "123", "1235", "567", "88"]
    phone_book4 = ["113333", "115555", "345555", "555555", "345444"]

    result1 = False
    result2 = True
    result3 = False
    result4 = True

    if(solution(phone_book1) == result1): print("RIGHT1")
    if(solution(phone_book2) == result2): print("RIGHT2")
    if(solution(phone_book3) == result3): print("RIGHT3")
    if(solution(phone_book4) == result4): print("RIGHT4")

if __name__ == "__main__":
    main()


'''
# 다른사람이 쓴 매우 간결한 코드
# startswith() 함수 사용
def solution(phoneBook):
    phoneBook = sorted(phoneBook)

    for p1, p2 in zip(phoneBook, phoneBook[1:]):
        if p2.startswith(p1):
            return False
    return True
'''